packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +541,672 @@ class ShellRoute extends RouteBase {
+/// where the List of Widgets represent the Navigators for each branch.
+///
+/// Example:
+/// ```
+/// builder: (BuildContext context, StatefulShellRouteState state, Widget child) {
+///   return StatefulNavigationShell(
+///     shellRouteState: state,
+///     containerBuilder: (_, __, List<Widget> children) => MyCustomShell(shellState: state, children: children),
+///   );
+/// }
+/// ```
+class StatefulNavigationShell extends StatefulWidget {
+  /// Constructs an [_StatefulNavigationShell].
+  const StatefulNavigationShell(
+      {required this.shellRouteState,
+      ShellNavigatorContainerBuilder? containerBuilder,
+      super.key})
+      : containerBuilder = containerBuilder ?? _defaultChildBuilder;
+
+  static Widget _defaultChildBuilder(BuildContext context,
+      StatefulShellRouteState shellRouteState, List<Widget> children) {
+    return _IndexedStackedRouteBranchContainer(
+        currentIndex: shellRouteState.currentIndex, children: children);
+  }
+
+  /// The current state of the associated [StatefulShellRoute].
+  final StatefulShellRouteState shellRouteState;
+
+  /// The builder for a custom container for shell route Navigators.
+  final ShellNavigatorContainerBuilder containerBuilder;
+
+  @override
+  State<StatefulWidget> createState() => _StatefulNavigationShellState();
+}
+
+/// State for StatefulNavigationShell.
+class _StatefulNavigationShellState extends State<StatefulNavigationShell> {
+  final Map<Key, Widget> _navigatorCache = <Key, Widget>{};
+
+  final List<_StatefulShellBranchState> _branchStates =
+      <_StatefulShellBranchState>[];
+
+  StatefulShellRouteState get shellState => widget.shellRouteState;
+
+  StatefulShellRoute get _route => widget.shellRouteState.route;
+
+  Widget? _navigatorForBranch(StatefulShellBranch branch) {
+    return _navigatorCache[branch.navigatorKey];
+  }
+
+  void _setNavigatorForBranch(StatefulShellBranch branch, Widget? navigator) {
+    navigator != null
+        ? _navigatorCache[branch.navigatorKey] = navigator
+        : _navigatorCache.remove(branch.navigatorKey);
+  }
+
+  void _switchBranch(int index) {
+    final GoRouter goRouter = GoRouter.of(context);
+    final _StatefulShellBranchState branchState = _branchStates[index];
+    final RouteMatchList? matchList =
+        branchState.matchList?.modifiableMatchList;
+    if (matchList != null) {
+      goRouter.routeInformationParser
+          .processRedirection(matchList, context)
+          .then(goRouter.routerDelegate.setNewRoutePath)
+          .onError((_, __) =>
+              goRouter.go(_defaultBranchLocation(branchState.branch)));
+    } else {
+      goRouter.go(_defaultBranchLocation(branchState.branch));
+    }
+  }
+
+  String _defaultBranchLocation(StatefulShellBranch branch) {
+    return branch.initialLocation ??
+        GoRouter.of(context)
+            .routeConfiguration
+            .findStatefulShellBranchDefaultLocation(branch);
```

Comment:
I'm not sure it can be separated out of RouteConfiguration really, but I've anyway perhaps found a better solution, where I've added a goBranch to GoRouter, to avoid exposing/accessing RouteConfiguration from other areas. Stay tuned for commits.